orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java

```diff
@@ -72,6 +74,63 @@ private void replaceStages(PipelineTemplate pipelineTemplate) {
     }
   }
 
+  private void expandStagePartials(PipelineTemplate pipelineTemplate) {
+    List<StageDefinition> templateStages = pipelineTemplate.getStages();
+
+    List<StageDefinition> addStages = new ArrayList<>();
+    templateStages.stream().filter(StageDefinition::isPartialType).forEach(partialPlaceholder -> {
+      PartialDefinition partial = pipelineTemplate.getPartials().stream()
+        .filter(p -> p.getRenderedPartials().containsKey(partialPlaceholder.getId()))
+        .findFirst()
+        .orElseThrow(() -> new IllegalTemplateConfigurationException(String.format("Could not find rendered partial: %s", partialPlaceholder.getId())));
+
+      List<StageDefinition> stages = partial.getRenderedPartials().get(partialPlaceholder.getId());
+
+      // Create a graph so we can find the leaf nodes
+      createGraph(stages);
+
+      // get leaf nodes of the partial
+      Map<String, Integer> outOrder = new HashMap<>();
+      Set<StageDefinition> sorted = new LinkedHashSet<>();
+      Map<String, Status> state = new HashMap<>();
+
+      Map<String, StageDefinition> graph = stages
+        .stream()
+        .collect(
+          Collectors.toMap(StageDefinition::getId, i -> i)
+        );
+
+      graph.keySet().forEach(k -> outOrder.put(k, 0));
+      graph.forEach((k, v) -> dfs(k, sorted, state, graph, outOrder));
+      Set<String> leafNodes = sorted
+        .stream()
+        .filter(i -> outOrder.get(i.getId()) == 0)
+        .map(StageDefinition::getId)
+        .collect(Collectors.toSet());
+
+      // assign root nodes to placeholder's dependsOn value
+      stages.stream()
```

Comment:
That's correct. I think I'll break this method up a little bit so it's more descriptive of what's going, but essentially:

`stage1 -> partial(stage3, stage4) -> stage2`

needs to expand to:

`stage1 -> stage3 -> stage4 -> stage2`

So, I build the graph, then assign stage3's dependsOn to the value of `partial1`, then stage2's dependsOn to the leaf nodes of the partial's internal graph. Once I've done those operations, I add the partial's stages into the template and remove the `partial1` stage from the graph. Make sense?